plutus-core/plutus-ir/src/PlutusIR/Purity.hs

```diff
@@ -142,6 +147,20 @@ termEvaluationOrder semvar varStrictness = goTerm
 
         -- If we can view as a builtin application, then handle that specially
         (splitApplication -> (Builtin a fun, args)) -> goBuiltinApp a fun args
+        -- If we can view as a constructor application, then handle that specially.
+        -- Constructor applications are always pure: if under-applied they don't
+        -- reduce; if fully-applied they are pure; if over-applied it's going to be
+        -- a type error since they never return a function. So we can ignore the arity
+        -- in this case!
+        (splitApplication -> (h@(Var _ n), args))
+          | Just (DatatypeConstructor{}) <- lookupVarInfo n vinfo ->
+            evalThis (EvalTerm Pure MaybeWork h)
+            <>
+            goAppCtx args
```

Comment:
Aren't the args evaluated before `this`?

Reply:
`h` is the head of the application, applications are still evaluated left-to-right, so the head is always first.